### PR TITLE
refactor(ai): align prompt template and chat APIs

### PR DESCRIPTION
**Pull request by andrewheard**
Not ready for review, needs more work.
TODO:
- Add support for function calling
- Add CHANGELOG entry
- Update integration tests and add one for function calling

Updated the Server Prompt Template APIs in `FirebaseAI` to align with the core `GenerativeModel` and `Chat` API design, brought parity with the Android/Kotlin SDK equivalents (`TemplateGenerativeModel` and `TemplateChat`), and graduated prompt templates from public preview to a fully public API.

### 1. API Alignment with `GenerativeModel` / `Chat` and Android
- **Instance-level configuration**: Moved `RequestOptions` and `TemplateToolConfig` from per-request parameters on `generateContent` and `generateContentStream` to instance initialization via `FirebaseAI.templateGenerativeModel(toolConfig:requestOptions:)`, matching `GenerativeModel` and Android's constructor.
- **Session-scoped template inputs**: Scoped template variables (`inputs: [String: Any]`) to session initialization via `startChat(templateID:inputs:history:)` rather than requiring them on every message send, matching `TemplateChat` on Android.
- **Streamlined message sending**: Simplified `TemplateChatSession.sendMessage` and `sendMessageStream` to only accept message content, added variadic `PartsRepresentable...` overloads that match `Chat`, and handled stream task cancellation.

### 2. Public API Promotion
- Graduated prompt templates from preview status by removing public preview disclaimers across `TemplateGenerativeModel` and `TemplateChatSession`.
- Made `TemplateChatSession`, its streaming methods, and `TemplateGenerativeModel.startChat` `public`.

### 3. Documentation & Tests
- Updated and corrected DocC comments to reflect parameter changes, documented `TemplateChatSession.history`, and removed obsolete parameter references.
- Updated unit tests in `TemplateChatSessionTests` and `TemplateGenerativeModelTests` to reflect the streamlined initialization and session workflows.

#no-changelog

### CODE REVIEW

File: FirebaseAI/Sources/TemplateChat.swift [hunk 1]
```diff
@@ -41,81 +61,53 @@ final class TemplateChatSession: Sendable {
...
-  func sendMessage(_ message: any PartsRepresentable,
-                   inputs: [String: Any],
-                   options: RequestOptions = RequestOptions()) async throws
+  public func sendMessage(_ parts: any PartsRepresentable...) async throws
     -> GenerateContentResponse {
-    return try await sendMessage([ModelContent(parts: message.partsValue)],
-                                 inputs: inputs,
-                                 options: options)
+    return try await sendMessage([ModelContent(parts: parts)])
   }
```
[the remark this answers is not in the document]
Comment (andrewheard) [reply FirebaseAI/Sources/TemplateChat.swift:112]: Addressed in [`51dd665`](https://github.com/firebase/firebase-ios-sdk/pull/16676/commits/51dd665ba52c658def21a7fcaaf29d7e1e078482) and updated the integration test to use `chat.sendMessage()`, which verifies against the backend that this edge case is handled.

File: FirebaseAI/Sources/TemplateChat.swift [hunk 2]
```diff
@@ -142,31 +134,25 @@ final class TemplateChatSession: Sendable {
...
-  func sendMessageStream(_ message: any PartsRepresentable,
-                         inputs: [String: Any],
-                         options: RequestOptions = RequestOptions()) throws
+  public func sendMessageStream(_ parts: any PartsRepresentable...) throws
     -> AsyncThrowingStream<GenerateContentResponse, Error> {
-    return try sendMessageStream([ModelContent(parts: message.partsValue)],
-                                 inputs: inputs,
-                                 options: options)
+    return try sendMessageStream([ModelContent(parts: parts)])
   }
```
[the remark this answers is not in the document]
Comment (andrewheard) [reply FirebaseAI/Sources/TemplateChat.swift:183]: Addressed in [`51dd665`](https://github.com/firebase/firebase-ios-sdk/pull/16676/commits/51dd665ba52c658def21a7fcaaf29d7e1e078482) and updated the integration test to use `chat.sendMessage()`, which verifies against the backend that this edge case is handled.

File: FirebaseAI/Sources/Types/Internal/Tools/TemplateFunction.swift [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        EncodingError.Context(
+          codingPath: [],
+          debugDescription: "Server Prompt Templates do not support automatic function calling."
+        )
+      )
+    }
+
+    name = functionDeclaration.name
+
+    if let parameters = functionDeclaration.parameters {
```
Comment (paulb777) [inline FirebaseAI/Sources/Types/Internal/Tools/TemplateFunction.swift:83]: ## Dead code / missing capability

### 3. `TemplateFunction.init(_:)` — unreachable branches, `outputSchema` can never be set

[`TemplateFunction.swift`](https://github.com/firebase/firebase-ios-sdk/pull/16676/files) guards on `.manual`:

```swift
guard case .manual = functionDeclaration.kind else { throw ... }
...
if let parameters = functionDeclaration.parameters {
  inputSchema = parameters.toJSONSchema()
} else if let parametersJSONSchema = functionDeclaration.parametersJSONSchema {
  inputSchema = try parametersJSONSchema.toGeminiJSONSchema()   // unreachable
} else {
  inputSchema = nil                                             // unreachable
}
outputSchema = functionDeclaration.responseJSONSchema           // always nil
```

In `FunctionDeclaration`'s only public initializer (`Tool.swift`), `kind = .manual` always sets `parameters` non-nil and `parametersJSONSchema = responseJSONSchema = nil`. The `.foundationModels` path is the only one that populates those, and it's rejected by the guard.

So:
- The `else if` and `else` branches are unreachable, and `init` can't actually throw after the guard.
- **`outputSchema` is permanently `nil`** — confirmed by the PR's own assertion `XCTAssertNil(functions[0].outputSchema)`.

Two things to decide:
1. Is it intentional that clients cannot supply a function **output** schema to a template? If the backend supports it, this is a missing feature; if not, consider dropping `outputSchema` from `TemplateFunction` rather than shipping a field that's always absent.
2. Either simplify to `inputSchema = functionDeclaration.parameters?.toJSONSchema()`, or keep the defensive branches with a comment explaining they're future-proofing against the guard being relaxed.

File: FirebaseAI/Sources/TemplateChat.swift [hunk 4]
```diff
@@ -0,0 +1,194 @@
+// Copyright 2025 Google LLC
```
Comment (paulb777) [inline FirebaseAI/Sources/TemplateChat.swift:1]:
```suggestion
// Copyright 2026 Google LLC
```

**Review by paulb777 — APPROVED** [review]
LGTM

File: FirebaseAI/Sources/TemplateChat.swift (hunk 4 above)
Comment (andrewheard) [reply FirebaseAI/Sources/TemplateChat.swift:1]: This one was my fault since I didn't notice it wasn't marked as a rename. The original code is in https://github.com/firebase/firebase-ios-sdk/pull/16676/changes/BASE..92c524434a5123b6a01ec938d73e359dd5f985e5#diff-cc84b3e977b2e9598002aabad2518ea174c65631b8340ac1c1e15839ef293d2d.

Let me know if we should still switch it to 2026, easy fix.

Comment (paulb777) [reply FirebaseAI/Sources/TemplateChat.swift:1]: not necessary for a rename

File: FirebaseAI/Sources/Types/Internal/Tools/TemplateFunction.swift (hunk 3 above)
Comment (andrewheard) [reply FirebaseAI/Sources/Types/Internal/Tools/TemplateFunction.swift:83]: This is something we currently only support in the Foundation Models integration. The current approach we offer for creating a function declaration is: https://github.com/firebase/firebase-ios-sdk/blob/f55af5b56f2edc781fd2dba01eda1f193c8f0a1f/FirebaseAI/Sources/Tool.swift#L46-L67

Since it accepts a `Schema`, for `parameters`, in OpenAPI 3.0 schema format, this PR converts `Schema` under the hood to JSON Schema format in [`toJSONSchema()`](https://github.com/firebase/firebase-ios-sdk/pull/16676/changes/BASE..92c524434a5123b6a01ec938d73e359dd5f985e5#diff-ee64a8f73bed7a657eeed9f0a4537a3606b7fbf0d44ceceafac9390405523eb1). I couldn't think of another way to do it on short notice without introducing another new public API.

Note: All of the unreachable code will be cleaned up once I remove the pre-`GeminiLanguageModel` integration with Foundation Models in favour of the native integration using `LanguageModel`.
